### ISSUE TITLE
I'm back baby! 

### ISSUE THREAD

**Issue by crashkonijn** [crashkonijn/GOAP#8]
Hi all,

First of all I'd like to thank everyone that has taken an interest to this project over the years!

After doing this as my graduation project back in 2016 I unfortunately didn't end up working in games and with that I didn't have the energy to put more effort into this project.

The last year I've started doing some private game dev to get back into the game 🕹 and for my latest project I started rebuilding a GOAP from scratch! Luckily I have been working as a .NET developer so my relevant coding experience did increase ✌

So far the project is composed of two separate projects:

- A "Pure" GOAP written in C#. It knows nothing of unity and is only the node navigator in it's purest sense.
- The Unity implementation of the above mentioned package including debug windows, editors and configuration through scriptable objects!

# Features
So this is what I've got going so far:

- [x] Goals
- [x] Actions
- [x] Sensors (A proper example was lacking in the 2016 version)
- [x] Easy setup through scriptable objects
- [x] Proper abstractions
- [x] Proper demo
- [x] Basic versions of debug windows
- [x] Tests (partially)
- [ ] Packaging it as a unity package
- [ ] Documentation
- [ ] Running tests in a pipeline

Notes:
- The debug windows are in somewhat rough state still (code wise) and still require some proper cleaning
- It's not multithreaded (for now)

# Help
If you'd like to get involved in the development or can help with the packaging or pipeline please let me know!

# Sneak preview:
![image](https://user-images.githubusercontent.com/5364802/202007995-f812a159-8075-49fc-b7e0-b3f7fb7371da.png)
![image](https://user-images.githubusercontent.com/5364802/202008297-0d248ecb-e390-4873-ba9d-45954e78590e.png)
![image](https://user-images.githubusercontent.com/5364802/202011157-1a3467e4-d5ba-407a-b6fb-79d50034e5f2.png)
![image](https://user-images.githubusercontent.com/5364802/202008374-e0a4c81b-032d-4cb2-ba45-886f4699ca3e.png)
![image](https://user-images.githubusercontent.com/5364802/202007932-e33e3f79-907d-4146-b570-183da25ba98f.png)
![image](https://user-images.githubusercontent.com/5364802/202008652-70e3479c-f7d6-44bf-b532-45dcb27315c6.png)

**Comment by ice-ko** [comment]
Welcome back, looking forward to the follow-up updates,

**Comment by crashkonijn** [comment]
Thanks!

I've had a bit of a rough December, but I'll be continuing soon!

Btw, #9 has a much better overview of the work being done.

**Comment by ice-ko** [comment]
hi crashkonijn,
Can you make a tutorial video for the feature/version-2 branch? I don't understand something,
Thanks!

**Comment by crashkonijn** [comment]
Hi!

Have you had a look at the unity project? I'm aware that it might be a bit simple as an example.

I don't prefer writing docs/recording a vid untill the code has settled a bit more.

Could you tell me what you don't understand? I could explain it or try and make it more obvious.

Thanks!

**Comment by ice-ko** [comment]
> Hi!
> 
> Have you had a look at the unity project? I'm aware that it might be a bit simple as an example.
> 
> I don't prefer writing docs/recording a vid untill the code has settled a bit more.
> 
> Could you tell me what you don't understand? I could explain it or try and make it more obvious.
> 
> Thanks!

I saw your demo, and I can’t see the information nodes when I open the GOAP Editor window. Is it because the update is not perfect?

**Comment by ice-ko** [comment]
I want to join the task system, such as: cutting trees, picking fruits, planting, etc. Do you have any better suggestions?

**Comment by crashkonijn** [comment]
Hey @ice-ko 

I've just done a minor update which makes the editor windows more accessible (Tools -> Goap).

Please note that with all the refactoring I did lately the `Setup Inspector` window is now broken. The `Agents Node` window should work during play mode (when selecting an agent).

I'm still planning on updating all the editor windows, however I'm still thinking whether or not I should use 2022.2 for that (with the improved UI Toolkit).

Please do note that this package is in active development and anything might still change. In the near future I'm planning on integrating this project in one of my own projects, that should give me better insight in what does or doesn't work and still needs improving.

> I want to join the task system, such as: cutting trees, picking fruits, planting, etc. Do you have any better suggestions?

I'm not quite sure what you're asking here. Could you rephrase your question?

**Comment by ice-ko** [comment]
> Hey @ice-ko
> 
> I've just done a minor update which makes the editor windows more accessible (Tools -> Goap).
> 
> Please note that with all the refactoring I did lately the window is now broken. The window should work during play mode (when selecting an agent).`Setup Inspector``Agents Node`
> 
> I'm still planning on updating all the editor windows, however I'm still thinking whether or not I should use 2022.2 for that (with the improved UI Toolkit).
> 
> Please do note that this package is in active development and anything might still change. In the near future I'm planning on integrating this project in one of my own projects, that should give me better insight in what does or doesn't work and still needs improving.
> 
> > I want to join the task system, such as: cutting trees, picking fruits, planting, etc. Do you have any better suggestions?
> 
> I'm not quite sure what you're asking here. Could you rephrase your question?

ok i'll check for an update,
ai task system, assign tasks to ai to perform, such as task systems such as construction, collection, and felling in RimWorld

**Comment by ice-ko** [comment]
> Hey @ice-ko
> 
> I've just done a minor update which makes the editor windows more accessible (Tools -> Goap).
> 
> Please note that with all the refactoring I did lately the `Setup Inspector` window is now broken. The `Agents Node` window should work during play mode (when selecting an agent).
> 
> I'm still planning on updating all the editor windows, however I'm still thinking whether or not I should use 2022.2 for that (with the improved UI Toolkit).
> 
> Please do note that this package is in active development and anything might still change. In the near future I'm planning on integrating this project in one of my own projects, that should give me better insight in what does or doesn't work and still needs improving.
> 
> > I want to join the task system, such as: cutting trees, picking fruits, planting, etc. Do you have any better suggestions?
> 
> I'm not quite sure what you're asking here. Could you rephrase your question?

ok i'll check for an update,
ai task system, assign tasks to ai to perform, such as task systems such as construction, collection, and felling in RimWorld

**Comment by crashkonijn** [comment]
Hi @ice-ko

Thanks for the clarification!

This package only focusses on the core elements of GOAP. You can specify goals and actions and it will figure out the best action to reach that goal. Performing and switching to the next action is something that this package will handle for you.

Something that it doesn't do is determining what the current goal is for an agent. (Which is something the original version does do, but was very limited to control). Since determining the goal is something that is very game specific I don't provide such a feature.

Changing a goal for an agent is all up to you, however it's very easy to do. An example can be seen in the [AgentBrain](https://github.com/crashkonijn/GOAP/blob/feature/version-2/Demo/Assets/Demos/Behaviours/AgentBrain.cs) script in the demo.

Right now the example is a bit basic, but I will extend that in the future.

If you have any further questions, let me know 😄

**Comment by crashkonijn** [comment]
Closing this issue, I just released the rebuild of this project!
